src/Uart/Uart.c

```diff
@@ -0,0 +1,448 @@
+        if(ByteFifo_isEmpty(uart->txFifo) == true
+           && (uart->reg->status & UART_STATUS_TS) != 0) {
+            return true;
+        }
+    }
+    ByteFifo_clear(uart->txFifo);
+    *errCode = Uart_ErrorCode_Timeout;
+    return false;
+}
+
+bool
+Uart_read(Uart* const uart,
+          uint8_t* data,
+          uint32_t const timeoutLimit,
+          Uart_ErrorCode* const errCode)
+{
+    uint8_t byte = '\0';
+    uint32_t timeout = timeoutLimit;
+    do {
+        if(errorFlags[uart->id].hasRxFifoFullErrorOccurred == true) {
+            *errCode = Uart_ErrorCode_RxFifoFull;
+            return false;
+        }
+        if(ByteFifo_isEmpty(uart->rxFifo) == false) {
+            ByteFifo_pull(uart->rxFifo, &byte);
+            *data = byte;
+            return true;
+        }
+    } while((timeoutLimit == 0) || timeout-- > 0);
+    ByteFifo_clear(uart->rxFifo);
+    *errCode = Uart_ErrorCode_Timeout;
+    return false;
+}
+
+void
+Uart_writeAsync(Uart* const uart,
+                ByteFifo* const fifo,
+                const Uart_TxHandler handler)
+{
+    rtems_interrupt_vector_disable(interruptNumber(uart->id));
+    uart->txFifo = fifo;
+    uart->txHandler = handler;
+    uint8_t byte = '\0';
+    if((uart->reg->status & UART_STATUS_TE) != 0) {
+        ByteFifo_pull(uart->txFifo, &byte);
+        uart->reg->data = byte;
+    }
+    rtems_interrupt_vector_enable(interruptNumber(uart->id));
+}
+
+void
+Uart_readAsync(Uart* const uart,
+               ByteFifo* const fifo,
+               const Uart_RxHandler handler)
+{
+    rtems_interrupt_vector_disable(interruptNumber(uart->id));
+    uart->rxFifo = fifo;
+    uart->rxHandler = handler;
+    rtems_interrupt_vector_enable(interruptNumber(uart->id));
+}
+
+bool
+Uart_isTxEmpty(const Uart* const uart)
+{
+    return ByteFifo_isEmpty(uart->txFifo);
+}
+
+bool
+Uart_isRxEmpty(const Uart* const uart)
+{
+    return ByteFifo_isEmpty(uart->rxFifo);
+}
+
+uint32_t
+Uart_getTxFifoCount(Uart* const uart)
+{
+    return ByteFifo_getCount(uart->txFifo);
+}
+
+uint32_t
+Uart_getRxFifoCount(Uart* const uart)
+{
+    return ByteFifo_getCount(uart->rxFifo);
+}
+
+bool
+Uart_handleError(Uart* const uart)
+{
+    if(Uart_getLinkErrors(uart->reg->status, &(errorFlags[uart->id])) == true) {
+        uart->errorHandler.callback(uart->errorHandler.arg);
+        return true;
+    }
+    return false;
+}
+
+bool
+Uart_handleRx(Uart* const uart)
+{
+    uint8_t buf = '\0';
+    if((uart->reg->control & UART_CONTROL_RE) != 0
+       && (uart->reg->status & UART_STATUS_DR) != 0) {
+        buf = uart->reg->data;
+        if(ByteFifo_isFull(uart->rxFifo) == true) {
+            errorFlags[uart->id].hasRxFifoFullErrorOccurred = true;
+            return true;
+        }
+        if(buf == uart->rxHandler.targetCharacter) {
+            uart->rxHandler.characterCallback(uart->rxHandler.characterArg);
+        }
+        if(sentBytesCounter[uart->id] == 1) {
```

Comment:
?

Reply:
changed and now it is looking more clear